Remove beast::Thread (RIPD-1189):

scottschurr opened the pull request:
This pull request assumes that pull request #1954 or something like it survives code review. So ignore the first commit in this pull request. Assuming that #1954 is approved...

All uses of `beast::Thread` were previously removed from the code base, so `beast::Thread` is removed. One piece of `beast::Thread` needed to be preserved: the ability to set the current thread's name. So there's now a `beast::CurrentThreadName` that allows the current thread's name to be set and returned.

Thread naming is also cleaned up a bit. ThreadName.h and .cpp are removed since `beast::CurrentThreadName` does a better job. ThreadEntry is also removed, but its `terminateHandler()` is preserved in TerminateHandler.cpp. The revised `terminateHandler()` uses `beast::CurrentThreadName` to recover the name of the running thread.

Finally, the `NO_LOG_UNHANDLED_EXCEPTIONS` #define is removed since it was discovered that the MacOS debugger preserves the stack of the original throw even if the` terminateHandler()` rethrows.

In order to test the code on MacOS I thew an unhandled exception using both a debug and a release build. I did that both in a debugger and running from the command line. On MacOS these tests had good results. Both the debug and release builds wrote the `terminateHandler()` output to the terminal window and to the log. When the unhandled exception occurred inside the debugger I got both a full stack trace to the unhandled throw (not just to the `terminateHandler()`) and the output from the `terminateHandler()`. This in both the debug and the release builds.

I'm hoping that my reviewers, in addition to reviewing the code, can run similar tests on Linux and Windows. An easy way to cause an unhandled exception is to run the server after putting a `throw 5;` here: https://github.com/scottschurr/rippled/blob/rm-beast-thread/src/ripple/core/impl/Job.cpp#L83. But there are other choices. There may be a bit of back-and-forth as we get each platform to behave as well as possible.

Reviewers: @seelabs (Linux), @miguelportilla (Windows). Thanks.

nbougalis commented on src/ripple/app/ledger/impl/LedgerCleaner.cpp:
```diff
@@ -22,9 +22,8 @@
-#include <ripple/core/ThreadEntry.h>
 #include <ripple/protocol/JsonFields.h>
-#include <ripple/beast/core/Thread.h>
+#include <ripple/beast/core/CurrentThreadName.h>
```
I _think_ this fits better in `ripple/beast/utility`.

scottschurr replied:
How strongly do you feel about this one?  I did a quick run at this and I'm getting compile errors out of MacOS-specific headers.  My suspicion is that the core.unity.cpp file is doing some favors for CurrentThreadName.cpp. I can get that straightened out, but it feels like there would probably be corresponding problems for Windows and Linux that would take me a while to get under control.

scottschurr replied:
I dug into this a bit further.  On MacOS it looks like the files included by `beast/core/core.unity.cpp` (which at the moment includes `CurrentThreadName`) get special treatment so they can interoperate with Objective C.  It also looks like beast/core is the only place that gets this special treatment -- that's the only place I see `*.mm` files for Objective C.

So my inclination is to keep `CurrentThreadName` in `ripple/beast/core`, along with the other files that solve awkward OS integration problems.  Let me know what you think.  Thanks.

nbougalis replied:
I know this is settled, but I just wanted to add that I'm fine with what you have; thanks for investigating.

nbougalis commented on src/ripple/app/ledger/impl/LedgerCleaner.cpp:
```diff
@@ -243,13 +242,7 @@ class LedgerCleanerImp : public LedgerCleaner
-    void runImpl ()
-    {
-        beast::Thread::setCurrentThreadName ("LedgerCleaner");
+        beast::setCurrentThreadName ("LedgerCleaner");
```
I'd prefer `nameCurrentThread` or `setThreadName`.

seelabs commented:
NO_LOG_UNHANDLED_EXCEPTIONS needs to be removed from the CMakeFuncs.cmake

seelabs commented on src/ripple/beast/core/CurrentThreadName.cpp:
```diff
@@ -0,0 +1,149 @@
+
+std::string getCurrentThreadName ()
+{
+    return detail::threadName.get() ? *detail::threadName.get() : "";
```
Consider returning `boost::optional` (it simplifies callers). We're also calling `get` twice here. How about:
```
boost::optional<std::string> getCurrentThreadName ()
{
    if (auto r = detail::threadName.get())
        return *r;
    return boost::none;
}
```

seelabs replied:
How difficult would it be to get the thread name from the OS instead of a thread local? We're already registering the thread name with the OS? My guess is "it's a pain", in which case the thread local is fine.

seelabs commented on src/ripple/beast/core/CurrentThreadName.cpp:
```diff
@@ -0,0 +1,149 @@
+
+void setCurrentThreadNameImpl (std::string const& name)
+{
+   #if BEAST_DEBUG && BEAST_MSVC
```
Is `BEAST_WINDOWS` ever true without `BEAST_MSVC`? No change needed.

seelabs commented on src/ripple/beast/core/CurrentThreadName.cpp:
```diff
@@ -0,0 +1,149 @@
+
+    __try
+    {
+        RaiseException (0x406d1388 /*MS_VC_EXCEPTION*/, 0, sizeof (info) / sizeof (ULONG_PTR), (ULONG_PTR*) &info);
```
I do not understand the purpose of this code. It looks like we raise a structured exception and then catch it and continue (right?). I'm clearly missing something.

seelabs commented on src/ripple/beast/core/CurrentThreadName.cpp:
```diff
@@ -0,0 +1,149 @@
+
+namespace beast {
+
+void setCurrentThreadName (std::string name)
```
Parameter should be `const &`.

seelabs replied:
I missed the move; you're absolutely correct.

seelabs commented on src/ripple/beast/core/CurrentThreadName.cpp:
```diff
@@ -0,0 +1,149 @@
+namespace beast {
+namespace detail {
+
+static boost::thread_specific_ptr<std::string> threadName;
```
I thought all our platforms supported thread locals at this point? OTOH, we could change all the `boost::thread_specific_ptr` at once in it's own PR. No change needed.

HowardHinnant replied:
I think switching to `thread_local`  (globally) in a separate P/R is prudent.  macOS has only _very_ recently started supporting `thread_local`.

seelabs commented on src/ripple/core/impl/Job.cpp:
```diff
@@ -77,12 +77,15 @@ bool Job::shouldCancel () const
+
+    // Destroy the lambda, otherwise we won't include
+    // its duration in the time measurement
+    mJob = std::function<void(Job&)>();
```
`mJob = nullptr;` is terser and accomplishes the same thing.

seelabs commented on src/ripple/core/impl/TerminateHandler.cpp:
```diff
@@ -18,61 +18,54 @@
     if (std::current_exception())
     {
-        std::string const name = threadName.get() ? *threadName.get() : "Unknown";
+        std::string thread {beast::getCurrentThreadName()};
```
I don't expect a variable named `thread` to be of type `string`. Consider renaming to `name` or somesuch.

seelabs replied:
If we change `getCurrentThreadName` to return a `boost::optional`, we can change this to:
`auto const thread = beast::getCurrentThreadName().value_or("Unknown");`

HowardHinnant commented on src/ripple/beast/core/CurrentThreadName.cpp:
```diff
@@ -0,0 +1,149 @@
+    }
+    catch(...)
+    {
+    }
```
Just curious:  Why swallow exceptions here?  Was that our previous behavior?

scottschurr replied:
Yup, that code was teleported from here: https://github.com/ripple/rippled/blob/develop/src/ripple/core/impl/ThreadEntry.cpp#L35-L44.  You and Scott D. did that work together, so I assumed the idiom was correct.

seelabs replied:
I think it was for unhandled exception handling. We were worried about throwing an exception while another was in flight. (IIRC, but my memory is fuzzy).

scottschurr replied:
I'd be delighted to add a comment to the code if we're fairly confident Scott D. captured the motivation.  A couple of thumbs up on this review comment and I'll add a code comment.

seelabs replied:
I'm in favor of keeping the pokemon handler here, and adding a comment. One motivation for this code is to log unhandled exceptions. If code in a terminate handler also throws - err I'm actually not sure what happens. An infinite loop maybe? At any rate, it's easier to just swallow the exception.

Also, this code is very unlikely to throw and if it does the consequences of not noticing the failure are not significant.

HowardHinnant replied:
I don't really see a strong motivation to swallow exceptions here.  If an exception is thrown here it is likely because we're running out of memory, in which case we're already hosed.  However I don't feel strongly that this behavior needs to be changed here and now.  Might be wise to change that behavior in a separate P/R should we choose to do so.  I'm fine with a comment, not sure what it would say...  `// Questionable try/catch` ?

HowardHinnant replied:
Throwing from a terminate handler is a direct call to `abort()`.

HowardHinnant replied:
I don't mind keeping the `try/catch`.

seelabs replied:
I took another look at this. The case I was worried about - calling this function from the terminate handler - doesn't happen. And I don't see anther reason to keep it. It's safe to get rid of the try/catch block (@HowardHinnant is right as usual)

scottschurr replied:
`try/catch` removed.

miguelportilla commented:
Under the Visual Studio debugger, throwing where @scottschurr suggested produces about six dialogs reporting the exception.

miguelportilla commented:
As @scottschurr suggested, the multiple dialogs seem to be caused by the six job threads. After moving the  test throw to the resource manager, a single dialog appeared with a usable stack.

scottschurr commented:
Sorry folks.  The test coverage on this pull request was bothering me, so I added a couple of unit tests.  That means the unit tests need to be reviewed.  So I've removed the "Passed" label.  If I could get @miguelportilla or @seelabs to look over the most recent "[FOLD]" with the new unit tests I'd be grateful.  Thanks for the help.

nbougalis commented on src/ripple/app/main/Main.cpp:
```diff
@@ -190,7 +190,7 @@ int run (int argc, char** argv)
     using namespace std;
 
-    setCallingThreadName ("main");
+    beast::setCurrentThreadName ("main");
```
This would previously have set the thread's name to `rippled: main` but will now set it, only, to `main`.

scottschurr commented:
Thanks for catching the thread naming issue @nbougalis.  I have a [FOLD] commit that I believe addresses the issue.  I think the easiest way to verify the fix is to search for all calls to `setCallingThreadName()` in the old code.  All of those calls should now have a `"rippled: `" in front in the call to `beast::setCurrentThreadName()`.  It would be great if you could audit and let me know what you think.  Thanks.

scottschurr commented:
Squashed and rebased to 0.50.0.

scottschurr commented:
I'm removing the "Passed" label since a) this commit may be implicated in a unit test crash b) we can't remove `beast::Thread` yet while there's work being done on DeadlineTimer.

scottschurr commented:
Now that it looks like removing `beast::Thread` from `DeadlineTimer` did not directly cause a slowdown if ledger closures, we have good evidence that this pull request is also safe.  I've rebased to the latest develop and I'm marking the pull request as passed.

seelabs commented:
In 0.60.0-b7
